File: March-2023/C++/ReducingDishes.cxx
// Reducing Dishes
/* 3 Steps
Goal
Constraints
Choice
*/
class Solution {
public:
   
    int arr[501][501];
    int n;
    int solve(vector<int>& satisfaction, int time, int index){
        // Goal
        if(index>=n) return 0;
        // Memoization to prevent Timeout 
        if(arr[index][time]!=-1){
            return arr[index][time];
        }
        // Constraints and Choice
        int include = satisfaction[index]*time+solve(satisfaction,time+1,index+1);
        int exclude = solve(satisfaction,time,index+1);
        return arr[index][time]= max(include,exclude);
    }
    
    int maxSatisfaction(vector<int>& satisfaction) {
        memset(arr,-1,sizeof(arr));
        // sort to make sure the element at the end is multiplied with highest time for biggest value
        sort(satisfaction.begin(), satisfaction.end());
        n = satisfaction.size();
        return solve(satisfaction,1,0);
    }
};

File: March-2023/Java/ReducingDishes.java
// Reducing Dishes
/* 3 Steps
Goal
Constraints
Choice
*/
class Solution {
    int[][] arr = new int[501][501];
    public int solve(int[] satisfaction, int time, int index, int n){
        // Goal
        if(index==n) return 0;
        
        // Memoization
        if(arr[index][time]!=-1){
            return arr[index][time];
        }
        // Choice  & constraint
        int include =satisfaction[index]*time + solve(satisfaction,time+1,index+1,n);       
        int exclude = solve(satisfaction,time,index+1,n);
        return arr[index][time]=Math.max(include,exclude);        
    }
    public int maxSatisfaction(int[] satisfaction) {
        Arrays.sort(satisfaction);
        for(int[] a: arr){
        Arrays.fill(a,-1);
        }
        int n = satisfaction.length;
        return solve(satisfaction,1,0,n);
    }
}
